optuna/samplers/_qmc.py

```diff
@@ -0,0 +1,329 @@
+from collections import OrderedDict
+from typing import Any
+from typing import Dict
+from typing import Optional
+from typing import Sequence
+
+import numpy as np
+
+# We want to lazy import scipy.stats.qmc as it is slow to import.
+import scipy  # NOQA
```

Comment:
Could you use `try_import` and check in in the `__init__`?
e.g. https://github.com/optuna/optuna/blob/master/optuna/integration/tensorflow.py#L5

Reply:
Using `_LazyImport` looks reasonable. 🙋  (And maybe we can remove the top level `scipy` importing because `scipy.stats.qmc` is imported individually, right?)

Reply:
This PR contains several merge commits so syncing the master branch may not be intuitive.

In my environment, rebasing is done by the following:

```sh
# upstream	git@github.com:optuna/optuna.git (fetch)
# upstream	git@github.com:optuna/optuna.git (push)
> git checkout upstream master  # The name `upstream` refers to optuna/optuna.git

> git checkout feature/qmc-sampler

> git rebase -Xours origin/master

> git add .
> git rebase --continue
> git revert b9c1a0669698dbebc838783933f70ba440cb56e1
```

After rebasing, we will be able to use `_LazyImport` like that:

```
diff --git a/optuna/samplers/_qmc.py b/optuna/samplers/_qmc.py
index 8ab63535..9bc79190 100644
--- a/optuna/samplers/_qmc.py
+++ b/optuna/samplers/_qmc.py
@@ -13,6 +13,7 @@ import optuna
 from optuna import distributions
 from optuna import logging
 from optuna._experimental import experimental
+from optuna._imports import _LazyImport
 from optuna._transform import _SearchSpaceTransform
 from optuna.distributions import BaseDistribution
 from optuna.samplers import BaseSampler
@@ -287,17 +288,16 @@ class QMCSampler(BaseSampler):
         self._independent_sampler.after_trial(study, trial, state, values)
 
     def _sample_qmc(self, study: Study, search_space: Dict[str, BaseDistribution]) -> np.ndarray:
-
         # Lazy import because the `scipy.stats.qmc` is slow to import.
-        import scipy.stats.qmc
+        qmc_module = _LazyImport("scipy.stats.qmc")
 
         sample_id = self._find_sample_id(study, search_space)
         d = len(search_space)
 
         if self._qmc_type == "halton":
-            qmc_engine = scipy.stats.qmc.Halton(d, seed=self._seed, scramble=self._scramble)
+            qmc_engine = qmc_module.Halton(d, seed=self._seed, scramble=self._scramble)
         elif self._qmc_type == "sobol":
-            qmc_engine = scipy.stats.qmc.Sobol(d, seed=self._seed, scramble=self._scramble)
+            qmc_engine = qmc_module.Sobol(d, seed=self._seed, scramble=self._scramble)
         else:
             raise ValueError("Invalid `qmc_type`")
```

Reply:
Sorry, I misunderstood that scipy is not int the dependencies of Optuna, but it is.
https://github.com/optuna/optuna/blob/master/setup.py#L37

Please ignore this message.
And using _LazyImport looks good to me!